src/main/java/org/truffleruby/core/string/StringNodes.java

```diff
@@ -2451,17 +2430,40 @@ protected int getCacheLimit() {
 
     }
 
-    @CoreMethod(names = "upcase!", raiseIfFrozenSelf = true)
-    @ImportStatic(StringGuards.class)
-    public abstract static class UpcaseBangNode extends CoreMethodArrayArgumentsNode {
+    @NodeChild("bytes")
+    public abstract static class InvertAsciiCaseNode extends RubyNode {
+
+        @CompilationFinal private int lowerBound = -1;
+        @CompilationFinal private int upperBound = -1;
 
         @Child private RopeNodes.MakeLeafRopeNode makeLeafRopeNode = RopeNodes.MakeLeafRopeNode.create();
 
-        @Specialization(guards = "isSingleByteOptimizable(string)")
-        public DynamicObject upcaseSingleByte(DynamicObject string,
-                                              @Cached("create()") BranchProfile foundLowerCaseCharProfile,
-                                              @Cached("createBinaryProfile()") ConditionProfile noopProfile,
-                                              @Cached("create()") RopeNodes.BytesNode bytesNode) {
+        public static InvertAsciiCaseNode createLowerToUpper() {
+            final InvertAsciiCaseNode ret = StringNodesFactory.InvertAsciiCaseNodeGen.create(null);
+            ret.setBounds('a', 'z');
+
+            return ret;
+        }
+
+        public static InvertAsciiCaseNode createUpperToLower() {
+            final InvertAsciiCaseNode ret = StringNodesFactory.InvertAsciiCaseNodeGen.create(null);
+            ret.setBounds('A', 'Z');
```

Comment:
Why not make these 2 arguments part of the node's constructor?

Reply:
I couldn't figure out how to make that work with the DSL generated class. It seems to only set up constructor args for `@NodeChildren`. I could make them nodes, but that seemed really heavy-weight for something I just want to be a `final boolean`.